File: 04-2.rb
#!/usr/bin/env ruby

require 'numo/narray'

class Puzzle
  def initialize(input)
    @height = input.size
    @width = input[0].size
    @map = Numo::UInt8.zeros(@height, @width)
    @height.times do |y|
      @width.times do |x|
        @map[y, x] = input[y][x].codepoints[0]
      end
    end

    mas = [
      'M.S'.codepoints,
      '.A.'.codepoints,
      'M.S'.codepoints
    ]
    @mas_height = mas.size
    @mas_width = mas[0].size
    xmas = Numo::UInt8.zeros(3, 3)
    @mas_height.times do |y|
      @mas_width.times do |x|
        xmas[y, x] = mas[y][x]
      end
    end
    @xmases = [xmas, xmas.rot90(1), xmas.rot90(2), xmas.rot90(3)]
  end

  def match?(y, x)
    @xmases.any? do |xmas|
      res = true
      @mas_height.times do |my|
        @mas_width.times do |mx|
          next if xmas[my, mx] == 46 # .

          res = false if xmas[my, mx] != @map[y + my, x + mx]
        end
      end
      res
    end
  end

  def count_matches
    count = 0
    (@height - @mas_height + 1).times do |y|
      (@width - @mas_width + 1).times do |x|
        count += 1 if match?(y, x)
      end
    end
    count
  end
end

input = File.read('04.input').lines.map(&:strip)
puzzle = Puzzle.new(input)
puts puzzle.count_matches

File: 04-2.py
#!/usr/bin/env python

import numpy


class Puzzle:
    def __init__(self, input):
        self.height = len(input)
        self.width = len(input[0])
        self.map = numpy.zeros((self.height, self.width), dtype=numpy.uint8)
        for y in range(self.height):
            for x in range(self.width):
                self.map[y, x] = ord(input[y][x])

        mas = [
            [ord('M'), ord('.'), ord('S')],
            [ord('.'), ord('A'), ord('.')],
            [ord('M'), ord('.'), ord('S')],
        ]
        self.mas_height = len(mas)
        self.mas_width = len(mas[0])
        xmas = numpy.zeros((self.mas_height, self.mas_width), dtype=numpy.uint8)
        for y in range(self.mas_height):
            for x in range(self.mas_width):
                xmas[y, x] = mas[y][x]

        self.xmases = [xmas, numpy.rot90(xmas, 1), numpy.rot90(xmas, 2), numpy.rot90(xmas, 3)]

    def match(self, y, x):
        for xmas in self.xmases:
            res = True
            for my in range(self.mas_height):
                for mx in range(self.mas_width):
                    if xmas[my, mx] == 46:  # .
                        continue

                    if xmas[my, mx] != self.map[y + my, x + mx]:
                        res = False
            if res:
                return True

        return False

    def count_matches(self):
        count = 0
        for y in range(self.height - self.mas_height + 1):
            for x in range(self.width - self.mas_width + 1):
                if self.match(y, x):
                    count += 1
        return count


input = [row.strip() for row in open('04.input').readlines()]
puzzle = Puzzle(input)
print(puzzle.count_matches())
